### PR TITLE
Improve packet relay (event query)

### PR DESCRIPTION
**Pull request by siburu**

### CODE REVIEW

File: core/provers.go [hunk 1]
```diff
@@ -40,12 +41,14 @@ type IBCProvableQuerier interface {
 
 	// QueryChannelWithProof returns the Channel and its proof
 	QueryChannelWithProof(ctx QueryContext) (chanRes *chantypes.QueryChannelResponse, err error)
+}
 
-	// QueryPacketCommitmentWithProof returns the packet commitment and its proof
-	QueryPacketCommitmentWithProof(ctx QueryContext, seq uint64) (comRes *chantypes.QueryPacketCommitmentResponse, err error)
+type IBCPacketProver interface {
```
Comment (bluele) [inline core/provers.go:46]: NOTE: define a general prove function instead

File: chains/tendermint/query.go [hunk 2]
```diff
@@ -145,50 +145,64 @@ func (c *Chain) QueryDenomTraces(ctx core.QueryContext, offset, limit uint64) (*
...
-		return nil, fmt.Errorf("can't find the packet from events")
+		return nil, clienttypes.Height{}, fmt.Errorf("can't find the packet from events")
 	}
-	return ack.Data(), nil
-}
 
-// QueryPacketReciept returns the packet reciept proof at a given height
-func (c *Chain) QueryPacketReciept(ctx core.QueryContext, seq uint64) (recRes *chantypes.QueryPacketReceiptResponse, err error) {
-	return chanutils.QueryPacketReceipt(c.CLIContext(int64(ctx.Height().GetRevisionHeight())), c.PathEnd.PortID, c.PathEnd.ChannelID, seq, true)
+	height := clienttypes.NewHeight(0, uint64(txs[0].Height))
```
Comment (bluele) [inline chains/tendermint/query.go:203]: Can't we set a non-zero revision number here?

File: core/provers.go (hunk 1 above)
Comment (siburu) [reply core/provers.go]: Fixed: https://github.com/hyperledger-labs/yui-relayer/pull/90/commits/dac8b52d62948506725691eca74ac4620834c129#diff-97524160c10dacd27d0cfdf4415d7c72cf4f84d7779893a6173aad7dbab270f4R46-R55

File: chains/tendermint/query.go (hunk 2 above)
Comment (siburu) [reply chains/tendermint/query.go]: Fixed: https://github.com/hyperledger-labs/yui-relayer/pull/90/commits/04d77bcdd90a1e6f6385fe46d33f126e86025b4a#diff-b9642f5116a057732bc245b582fcd98b7dcd3a6f0ceb83789ad6116f3da4b778R300

File: chains/tendermint/query.go [hunk 3]
```diff
@@ -249,25 +238,125 @@ func (c *Chain) QueryUnrecievedAcknowledgements(ctx core.QueryContext, seqs []ui
...
+		packet, rpHeight, err := c.queryReceivedPacket(ctx, ps.Sequence)
+		if err != nil {
+			return nil, err
+		}
+		ack, waHeight, err := c.queryWrittenAcknowledgement(ctx, ps.Sequence)
+		if err != nil {
+			return nil, err
+		}
+		var height clienttypes.Height
+		if rpHeight.LT(waHeight) {
```
Comment (bluele) [inline chains/tendermint/query.go:258]: Can we rewrite the following?
```
packets = append(packets, &core.PacketInfo{
	Packet:          *packet,
	Acknowledgement: ack,
	EventHeight:     rpHeight,
})
```
Considering the order of receive-packet and write-acknowledgment, `rpHeight <= waHeight` always holds.
Therefore, if `min(rpHeight, waHeight)` is assigned to `EventHeight`, then `EventHeight` will always be equal to `rpHeight`. Do I misunderstand something?

File: core/types.go [hunk 4]
```diff
@@ -1,7 +1,45 @@
...
+// a pair of RecvPacket/WriteAcknowledgement events. In the former case, the `Acknowledgement`
+// field becomes nil. In the latter case, `EventHeight` represents the smaller height between
+// the ones of the `RecvPacket` event and the `WriteAcknowledgement` event.
+type PacketInfo struct {
+	chantypes.Packet
+	Acknowledgement []byte             `json:"acknowledgement"`
+	EventHeight     clienttypes.Height `json:"event_height"`
+}
+
+type PacketInfoList []*PacketInfo
```
Comment (bluele) [inline core/types.go:18]: If you expect the elements to be ordered in ascending order of packet sequence, could you write this as a comment?

Comment (siburu) [reply core/types.go:18]: No, `PacketInfoList` is sorted in the order in which events occur.
I will comment this. Thank you.

Comment (bluele) [reply core/types.go:20]: > No, PacketInfoList is sorted in the order in which events occur.
oh, sorry. You are right.

Comment (siburu) [reply core/types.go:20]: Fixed: https://github.com/hyperledger-labs/yui-relayer/pull/90/commits/9980e1375724fad9397ced3dce0fc855fd894e54

File: chains/tendermint/query.go (hunk 3 above)
Comment (siburu) [reply chains/tendermint/query.go]: Fixed: https://github.com/hyperledger-labs/yui-relayer/pull/90/commits/9980e1375724fad9397ced3dce0fc855fd894e54

**Review by bluele — APPROVED** [review]
LGTM 👍

**Pull request merged**